modules/build/src/main/scala/scala/build/Inputs.scala

```diff
@@ -271,22 +288,37 @@ object Inputs {
 
     assert(validElems.nonEmpty)
 
-    val (inferredWorkspace, inferredNeedsHash, workspaceOrigin) = validElems
-      .collectFirst {
-        case d: Directory => (d.path, true, WorkspaceOrigin.SourcePaths)
+    val (inferredWorkspace, inferredNeedsHash, workspaceOrigin) = {
+      val settingsFiles = projectSettingsFiles(validElems)
+      val dirsAndFiles = validElems.collect {
+        case d: Directory  => d
+        case f: SourceFile => f
       }
-      .getOrElse {
-        validElems.head match {
-          case elem: SourceFile => (elem.path / os.up, true, WorkspaceOrigin.SourcePaths)
-          case _: Virtual =>
-            val dir = homeWorkspace(validElems, directories)
-            (dir, false, WorkspaceOrigin.HomeDir)
-          case r: ResourceDirectory =>
-            // Makes us put .scala-build in a resource directory :/
-            (r.path, true, WorkspaceOrigin.ResourcePaths)
-          case _: Directory => sys.error("Can't happen")
-        }
+
+      settingsFiles.collectFirst { s =>
```

Comment:
That use of `collectFirst` looks odd. Wouldn't something like this work better?
```suggestion
      settingsFiles.headOption.map { s =>
```